File: src/main/java/com/martmists/ynet/blocks/CableBlock.java
package com.martmists.ynet.blocks;

import com.martmists.ynet.YNetMod;
import com.martmists.ynet.api.BaseProvider;
import net.minecraft.block.Block;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.block.ConnectingBlock;
import net.minecraft.item.ItemPlacementContext;
import net.minecraft.state.StateManager;
import net.minecraft.state.property.Property;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.BlockView;
import net.minecraft.world.IWorld;

public class CableBlock extends ConnectingBlock {
    public CableBlock(Settings settings) {
        super(0.1875F, settings);
        setDefaultState(getStateManager().getDefaultState()
                .with(NORTH, false)
                .with(EAST, false)
                .with(SOUTH, false)
                .with(WEST, false)
                .with(UP, false)
                .with(DOWN, false)
        );
    }

    @Override
    protected void appendProperties(StateManager.Builder<Block, BlockState> builder) {
        builder.add(NORTH, EAST, SOUTH, WEST, UP, DOWN);
    }

    @Override
    public BlockState getPlacementState(ItemPlacementContext ctx) {
        return this.withConnectionProperties(ctx.getWorld(), ctx.getBlockPos());
    }

    @Override
    public BlockState getStateForNeighborUpdate(BlockState state, Direction facing, BlockState neighborState, IWorld world, BlockPos pos, BlockPos neighborPos) {
        if (!state.canPlaceAt(world, pos)) {
            world.getBlockTickScheduler().schedule(pos, this, 1);
            return super.getStateForNeighborUpdate(state, facing, neighborState, world, pos, neighborPos);
        } else {
            Block block = neighborState.getBlock();
            return state.with(FACING_PROPERTIES.get(facing), block == this || block == YNetMod.CONNECTOR || block instanceof BaseProvider);
        }
    }

    public BlockState withConnectionProperties(BlockView world, BlockPos pos) {
        Block block = world.getBlockState(pos.down()).getBlock();
        Block block2 = world.getBlockState(pos.up()).getBlock();
        Block block3 = world.getBlockState(pos.north()).getBlock();
        Block block4 = world.getBlockState(pos.east()).getBlock();
        Block block5 = world.getBlockState(pos.south()).getBlock();
        Block block6 = world.getBlockState(pos.west()).getBlock();
        return this.getDefaultState()
                .with(DOWN, block == this || block == YNetMod.CONNECTOR)
                .with(UP, block2 == this || block2 == YNetMod.CONNECTOR)
                .with(NORTH, block3 == this || block3 == YNetMod.CONNECTOR)
                .with(EAST, block4 == this || block4 == YNetMod.CONNECTOR)
                .with(SOUTH, block5 == this || block5 == YNetMod.CONNECTOR)
                .with(WEST, block6 == this || block6 == YNetMod.CONNECTOR);
    }
}

File: src/main/kotlin/com/martmists/ynet/block/CableBlock.kt
package com.martmists.ynet.block

import com.martmists.ynet.YNetMod
import com.martmists.ynet.network.Network
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.ConnectingBlock
import net.minecraft.entity.LivingEntity
import net.minecraft.item.ItemPlacementContext
import net.minecraft.item.ItemStack
import net.minecraft.state.StateManager
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.BlockView
import net.minecraft.world.World
import net.minecraft.world.WorldAccess

class CableBlock(settings: Settings) : ConnectingBlock(0.1875f, settings) {
    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        builder.add(NORTH, EAST, SOUTH, WEST, UP, DOWN)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState? {
        return withConnectionProperties(ctx.world, ctx.blockPos)
    }

    override fun neighborUpdate(state: BlockState, world: World, pos: BlockPos, block: Block, fromPos: BlockPos, notify: Boolean) {
        super.neighborUpdate(state, world, pos, block, fromPos, notify)
        val x = withConnectionProperties(world, pos)
        if (state != x) {
            world.setBlockState(pos, x)
        }
    }

    override fun getStateForNeighborUpdate(state: BlockState, facing: Direction, neighborState: BlockState, world: WorldAccess, pos: BlockPos, neighborPos: BlockPos): BlockState {
        return if (!state.canPlaceAt(world, pos)) {
            world.blockTickScheduler.schedule(pos, this, 1)
            super.getStateForNeighborUpdate(state, facing, neighborState, world, pos, neighborPos)
        } else {
            state.with(FACING_PROPERTIES[facing], YNetMod.shouldConnectCable(world as World, pos))
        }
    }

    override fun onBroken(world: WorldAccess, pos: BlockPos, state: BlockState) {
        Network.removeCable(pos)
    }

    override fun onPlaced(world: World, pos: BlockPos, state: BlockState, placer: LivingEntity?, itemStack: ItemStack) {
        val tStart = System.nanoTime()
        Network.addCable(pos)
        val tEnd = System.nanoTime()
        println((tEnd - tStart).toString() + "ns")
    }

    fun withConnectionProperties(world: BlockView, pos: BlockPos): BlockState {
        return defaultState
                .with(DOWN, YNetMod.shouldConnectCable(world as World, pos.down()))
                .with(UP, YNetMod.shouldConnectCable(world, pos.up()))
                .with(NORTH, YNetMod.shouldConnectCable(world, pos.north()))
                .with(EAST, YNetMod.shouldConnectCable(world, pos.east()))
                .with(SOUTH, YNetMod.shouldConnectCable(world, pos.south()))
                .with(WEST, YNetMod.shouldConnectCable(world, pos.west()))
    }

    init {
        defaultState = getStateManager().defaultState
                .with(NORTH, false)
                .with(EAST, false)
                .with(SOUTH, false)
                .with(WEST, false)
                .with(UP, false)
                .with(DOWN, false)
    }
}